lib/Dialect/SystemC/SystemCTypes.cpp

```diff
@@ -101,4 +199,113 @@ void SystemCDialect::registerTypes() {
+  auto *ctxt = parser.getContext();
+
+  if (mnemonic == IntBaseType::getMnemonic()) {
+    type = IntBaseType::get(ctxt);
+    return success();
+  }
+  if (mnemonic == UIntBaseType::getMnemonic()) {
+    type = UIntBaseType::get(ctxt);
+    return success();
+  }
+  if (mnemonic == SignedType::getMnemonic()) {
+    type = SignedType::get(ctxt);
+    return success();
+  }
+  if (mnemonic == UnsignedType::getMnemonic()) {
+    type = UnsignedType::get(ctxt);
+    return success();
+  }
+  if (mnemonic == BitVectorBaseType::getMnemonic()) {
+    type = BitVectorBaseType::get(ctxt);
+    return success();
+  }
+  if (mnemonic == LogicVectorBaseType::getMnemonic()) {
+    type = LogicVectorBaseType::get(ctxt);
+    return success();
+  }
+  if (mnemonic == IntType::getMnemonic())
+    return parseIntegerOrBitVector<IntType>(parser, type);
+  if (mnemonic == UIntType::getMnemonic())
+    return parseIntegerOrBitVector<UIntType>(parser, type);
+  if (mnemonic == BigIntType::getMnemonic())
+    return parseIntegerOrBitVector<BigIntType>(parser, type);
+  if (mnemonic == BigUIntType::getMnemonic())
+    return parseIntegerOrBitVector<BigUIntType>(parser, type);
+  if (mnemonic == BitVectorType::getMnemonic())
+    return parseIntegerOrBitVector<BitVectorType>(parser, type);
+  if (mnemonic == LogicVectorType::getMnemonic())
+    return parseIntegerOrBitVector<LogicVectorType>(parser, type);
+
+  return failure();
+}
+
+static LogicalResult customTypePrinter(Type type, DialectAsmPrinter &printer) {
+  return TypeSwitch<Type, LogicalResult>(type)
+      .Case<IntType, UIntType, BigIntType, BigUIntType, BitVectorType,
+            LogicVectorType>([&](auto ty) {
+        printer << ty.getMnemonic() << "<" << ty.getWidth() << ">";
+        return success();
+      })
+      .Case<IntBaseType, UIntBaseType, SignedType, UnsignedType,
+            BitVectorBaseType, LogicVectorBaseType>([&](auto ty) {
+        printer << ty.getMnemonic();
+        return success();
+      })
+      .Default([](auto ty) { return failure(); });
+}
+
+/// Parse a type registered with this dialect.
+Type SystemCDialect::parseType(DialectAsmParser &parser) const {
+  Type type;
+  llvm::SMLoc loc = parser.getCurrentLocation();
+  StringRef mnemonic;
+
+  mlir::OptionalParseResult result =
+      generatedTypeParser(parser, &mnemonic, type);
+  if (result.has_value() && !result.value())
+    return type;
+
+  result = customTypeParser(parser, mnemonic, loc, type);
+  if (result.has_value() && !result.value())
+    return type;
+
+  parser.emitError(loc) << "unknown type `" << mnemonic
+                        << "` in dialect `systemc`";
+  return {};
+}
+
+/// Print a type registered with this dialect.
+void SystemCDialect::printType(Type type, DialectAsmPrinter &printer) const {
+  if (succeeded(generatedTypePrinter(type, printer)))
+    return;
+  if (succeeded(customTypePrinter(type, printer)))
+    return;
+  assert(false && "no printer for unknown `systemc` dialect type");
```

Comment:
The assert is only going to be in debug builds. What should this do in a release build? Is there a utility for generating an `<<UnknownType>>` string?

Reply:
If you let ODS auto-generate this function, it won't even insert an assert. I don't think there's such a utility, but I could insert a `printer << "<<UnknownType>>";`. However, a user should never be able to hit the assert as it means that a custom printer is not implemented but should have been (there is no generic format for types) and I think a developer should at least enable assertions.